File: language-support/js/daml-ledger/tests/arbitrary/get_package_response.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as jsc from 'jsverify';
import * as ledger from '../../src';

export const GetPackageResponse: jsc.Arbitrary<ledger.GetPackageResponse> =
    jsc.tuple([jsc.elements([ledger.HashFunction.SHA256]), jsc.string, jsc.string]).smap<ledger.GetPackageResponse>(
        ([hashFunction, hash, archivePayload]) => {
            return {
                hashFunction: hashFunction,
                hash: hash,
                archivePayload: archivePayload
            }
        },
        (request) => {
            return [request.hashFunction, request.hash, request.archivePayload];
        }
    );

File: canton-3x/community/ledger/ledger-json-api/src/main/scala/com/digitalasset/canton/http/admin/GetPackageResponse.scala
// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.canton.http.admin

import com.google.protobuf
import com.daml.ledger.api.v1.package_service

sealed abstract class HashFunction extends Product with Serializable
case object SHA256 extends HashFunction
final case class Unrecognized(value: Int) extends HashFunction

object HashFunction {
  def fromLedgerApi(a: package_service.HashFunction): HashFunction = a match {
    case package_service.HashFunction.SHA256 => SHA256
    case package_service.HashFunction.Unrecognized(x) => Unrecognized(x)
  }
}

final case class GetPackageResponse(
    hashFunction: HashFunction,
    hash: String,
    archivePayload: protobuf.ByteString,
)

object GetPackageResponse {
  def fromLedgerApi(a: package_service.GetPackageResponse): GetPackageResponse = {
    GetPackageResponse(
      hashFunction = HashFunction.fromLedgerApi(a.hashFunction),
      archivePayload = a.archivePayload,
      hash = a.hash,
    )
  }
}
